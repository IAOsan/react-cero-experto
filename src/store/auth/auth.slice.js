import { createSlice } from '@reduxjs/toolkit';
import userService from '../../services/user.service';
import authService from '../../services/auth.service';
import {
	loadingStatus,
	successStatus,
	errorStatus,
	setError,
} from '../ui/ui.slice';

const slice = createSlice({
	name: '[auth]',
	initialState: {
		checking: true,
		isAuth: false,
		user: null,
	},
	reducers: {
		logged_in(state, { payload }) {
			state.checking = false;
			state.isAuth = true;
			state.user = payload;
		},
		logged_out(state) {
			state.isAuth = false;
			state.user = null;
		},
		checking_auth_started(state) {
			state.checking = true;
		},
		checking_auth_finished(state) {
			state.checking = false;
		},
	},
});

const { logged_in, logged_out, checking_auth_started, checking_auth_finished } =
	slice.actions;
export default slice.reducer;

/* -------------------------------------------------------------------------- */
/*                               action creators                              */
/* -------------------------------------------------------------------------- */
export const registerEmailAndPassword = (user) => async (dispatch) => {
	dispatch(loadingStatus());
	try {
		const res = await userService.register(user);
		dispatch(successStatus());
		dispatch(logged_in(res));
	} catch (error) {
		dispatch(errorStatus());
		dispatch(setError({ register: error }));
	}
};

export const loginEmailAndPassword = (user) => async (dispatch) => {
	dispatch(loadingStatus());
	try {
		const res = await authService.login(user);
		dispatch(successStatus());
		dispatch(logged_in(res));
	} catch (error) {
		dispatch(errorStatus());
		dispatch(setError({ login: error }));
	}
};

export const logout = () => (dispatch) => {
	try {
		authService.logout();
		dispatch(logged_out());
	} catch (error) {
		dispatch(setError(error));
	}
};

export const checkAuth = () => async (dispatch) => {
	const user = await authService.getCurrentUser();

	dispatch(checking_auth_started());
	if (user) {
		dispatch(logged_in(user));
	}
	dispatch(checking_auth_finished());
};

/* -------------------------------------------------------------------------- */
/*                                  selectors                                 */
/* -------------------------------------------------------------------------- */
export const selectAuthState = (state) => state.auth;
