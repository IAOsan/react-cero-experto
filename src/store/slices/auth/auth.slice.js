import { createSlice } from '@reduxjs/toolkit';
import {
	startLoading,
	finishLoading,
	addError,
	removeError,
} from '../ui/ui.slice';
import { clearNotes } from '../notes/notes.slice';
import * as authService from '../../../services/auth.service';

const initialState = {
	user: null,
	isAuth: false,
};

const slice = createSlice({
	name: '[auth]',
	initialState,
	reducers: {
		logged_in(state, { payload }) {
			state.user = payload;
			state.isAuth = true;
		},
		logged_out(state, action) {
			state.user = null;
			state.isAuth = false;
		},
	},
});

export default slice.reducer;

const { logged_in, logged_out } = slice.actions;

export const login = logged_in;

export const loginWithEmailPassWord = (data) => async (dispatch) => {
	dispatch(startLoading);
	dispatch(removeError);
	try {
		const user = await authService.loginEmailPassword(data);
		dispatch(logged_in(user));
	} catch (error) {
		dispatch(addError(error));
	} finally {
		dispatch(finishLoading);
	}
};

export const loginWithGoogle = async (dispatch) => {
	dispatch(removeError);
	try {
		const user = await authService.loginGoogle();
		dispatch(logged_in(user));
	} catch (error) {
		dispatch(addError(error));
	}
};

export const registerWithEmailPassword = (data) => async (dispatch) => {
	dispatch(startLoading);
	dispatch(removeError);
	try {
		const user = await authService.registerEmailPassword(data);
		dispatch(logged_in(user));
	} catch (error) {
		dispatch(addError(error));
	} finally {
		dispatch(finishLoading);
	}
};

export const logout = async (dispatch) => {
	await authService.logout();
	dispatch(logged_out());
	dispatch(clearNotes);
};
