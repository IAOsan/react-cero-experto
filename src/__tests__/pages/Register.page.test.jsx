import '@testing-library/jest-dom';
import { vi } from 'vitest';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
import { renderWithProviders, setupUser, clearAllAcounts } from '../test-utils';
import Register from '../../pages/Register.page';
import { registerWithEmailPassword } from '../../store/slices/auth/auth.slice';

vi.mock('../../store/slices/auth/auth.slice', async () => {
	const org = await vi.importActual('../../store/slices/auth/auth.slice');

	return {
		...org,
		registerWithEmailPassword: vi.fn(() => vi.fn()),
	};
});

const user = setupUser();

describe('<Register /> page', () => {
	beforeEach(async () => {
		await clearAllAcounts();
	});

	it('should renders correctly', () => {
		const wrapper = renderRegister();
		// snapshot
		expect(wrapper.container.firstElementChild).toMatchSnapshot();
		// register btn
		expect(
			wrapper.getByRole('button', { name: /register/i })
		).toBeInTheDocument();
		// footer to register
		expect(wrapper.getByText(/Already a user/i)).toBeInTheDocument();
		expect(
			wrapper.getByRole('link', { name: /login/i })
		).toBeInTheDocument();
	});

	it('should be able to register an user', async () => {
		const userData = {
			name: 'testUser',
			email: 'test@mail.com',
			password: '123456',
			confirmPassword: '123456',
		};

		const wrapper = renderRegister();

		await user.type(wrapper.getByPlaceholderText(/name/i), userData.name);
		await user.type(wrapper.getByPlaceholderText(/email/i), userData.email);
		await user.type(
			wrapper.getByPlaceholderText('Password'),
			userData.password
		);
		await user.type(
			wrapper.getByPlaceholderText('Confirm Password'),
			userData.confirmPassword
		);
		await user.keyboard('{Enter>1}');

		expect(registerWithEmailPassword).toHaveBeenCalledWith(userData);
	});

	it('should display an error if exists', () => {
		const error = {
			email: 'Access to this account has been temporarily disabled due to many failed login attempts',
		};
		const state = {
			ui: {
				error,
			},
		};
		const wrapper = renderRegister(['/auth/register'], state);

		expect(wrapper.getByText(error.email)).toBeInTheDocument();
	});

	it('should redirects to login page when clicks link', async () => {
		const wrapper = renderRegister();

		await user.click(wrapper.getByRole('link', { name: /LOGIN/i }));

		expect(wrapper.queryByRole('button', { name: /register/i })).toBeNull();
	});
});

function renderRegister(entries = ['/auth/register'], state = {}) {
	return renderWithProviders(
		<MemoryRouter initialEntries={entries}>
			<Routes>
				<Route path='/auth/register' element={<Register />} />
				<Route path='/auth/login' element={<mock-login />} />
			</Routes>
		</MemoryRouter>,
		{
			preloadedState: state,
		}
	);
}
