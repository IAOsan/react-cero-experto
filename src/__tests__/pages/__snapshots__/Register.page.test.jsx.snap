// Vitest Snapshot v1

exports[`<Register /> page > should renders correctly 1`] = `
<form
  class="form anima-fade-in-bck"
>
  <h2
    class="heading-2 mb-24 text-primary-alt"
  >
    Register
  </h2>
  <div
    class="form__group"
  >
    <input
      class="form__control"
      id="inputName"
      name="name"
      placeholder="Name"
      type="text"
      value=""
    />
  </div>
  <div
    class="form__group"
  >
    <input
      class="form__control"
      id="inputEmail"
      name="email"
      placeholder="Email"
      type="email"
      value=""
    />
  </div>
  <div
    class="form__group"
  >
    <input
      class="form__control"
      id="inputPassword"
      name="password"
      placeholder="Password"
      type="password"
      value=""
    />
  </div>
  <div
    class="form__group"
  >
    <input
      class="form__control"
      id="inputConfirmPassword"
      name="confirmPassword"
      placeholder="Confirm Password"
      type="password"
      value=""
    />
  </div>
  <button
    class="btn btn--default btn--primary btn--block my-20"
    disabled=""
    type="submit"
  >
    <div
      class="flex flex-ai-c flex-jc-c"
    >
      Register
    </div>
  </button>
  <p
    class="mt-20 text-center"
  >
    Already a user?
    <a
      class="mx-8 text-underline"
      href="/auth/login"
    >
      LOGIN
    </a>
  </p>
</form>
`;
